### ISSUE TITLE
Memory consumption + Speed

### ISSUE THREAD

**Issue by mastef** [kevinresol/hxgenjs#11]
This library is in general pretty fast!

I was wondering what would your suggestions be in terms of performance improvements, and where memory can be saved? I've been gradually testing `hxgenjs` by adding more and more classes from our main project, and I'm running out of 10 GB memory for projects that hit 10k classes. Ideally I'd like to run it on 30k+ classes ( most of them extend e.g. `MovieClip` or each other ).

In a project with 6k classes, the hxgenjs generator takes about : 
```
 28 seconds | to read in types
 90 seconds | generate 6068 types
 15 seconds | write 5449 types
----------------------------------
135 seconds total in `--times`
```

My initial attempts were : 
- to filter what ends up in the ClassProcessor cache ( only `id` and `externType` seem to be needed /re-used )
- changing DynamicAccess for stubs and typeIds to `Map<String, Bool>();` & `Map<TypeID, Bool>()`
- changing functions declarations inside methods to static functions, so they're not re-generated for every type
- moving `api` to a static public class, so it's not passed through everywhere

But with not much luck yet. Any other ideas?

**Comment by elsassph** [comment]
That doesn't seem wise to use hxgenjs for a large codebase - what is the goal?

**Comment by mastef** [comment]
The goal is to not output one massive `.js` file that has all 10-30k classes ( kills Chrome devtools, etc ), and to be able to bundle them in smaller chunks with js processing tools like webpack.

**Comment by kevinresol** [comment]
I reckon the memory problem comes from the fact that all type info (ProcessedClass, etc) are stored to the generate stage. Maybe they are using way to much memory. 
I did that because we need to generate the dependency graph. I'm not sure if it can be done with a "streaming" style,  which may help. 

Another option is to try using eval in haxe4. It may have better memory management.

**Comment by mastef** [comment]
Yeah tried a few approaches already. I think splitting the project is definitely an approach that seems good in theory currently. The only issue there being is handling common classes and libraries. 

Tried something similar with lime's [`-D modular` approach](https://github.com/mastef/openfl-Dmodular), which basically splits the project into multiple .hxml files, and then builds them separately while excluding the others with the `haxe.macro.Compiler.exclude()` macro, but [this little project with 5 classes split in 5 modules](https://github.com/mastef/openfl-Dmodular/tree/master/Source) unfortunately compiles in 60+ seconds on _each_ run.

Config : https://github.com/mastef/openfl-Dmodular/blob/master/project.xml
Resulting hxml files : https://github.com/mastef/openfl-Dmodular/tree/master/Export/html5/obj/lib ( see [line 20](https://github.com/mastef/openfl-Dmodular/blob/master/Export/html5/obj/lib/Mod1.hxml#L20) in each file, which calls [`ModuleHelper.exclude()`](https://github.com/openfl/lime/blob/939d73367f87088839a38d42231bc1e4909fbd7b/lime/tools/helpers/ModuleHelper.hx#L263-L272), which adds [`Compiler.exclude()`](https://github.com/HaxeFoundation/haxe/blob/3.4.4/std/haxe/macro/Compiler.hx#L242-L266) for each type )

I'm definitely looking for the "wise" solution 😅

A separat approach we're looking into is using `openfl-js` which moves the openfl library to externs only, which could reduce the exclusions for the openfl/lime libs, and just focus on the project.

@back2dos Do you think the `include` macro would be more performant than the `exclude` macro? How to handle class dependencies when going file-by-file?

**Comment by elsassph** [comment]
How do you obtain bundle splitting to such a hxgenjs project? Normally it requires an explicit dynamic import to trigger that.

I'd be curious how Webpack Haxe Loader / haxe-modular (not OpenFl's one) fare in comparison for huge codebase like that.

**Comment by mastef** [comment]
Linked to 2 example projects in the previous comment, but since I need OpenFL, they both use OpenFL.

**Comment by elsassph** [comment]
Ok I see - until this memory problem is resolved, I suppose (hopefully without offending anyone from `hxgenjs`) that https://github.com/jasononeil/webpack-haxe-loader would scale better for your use case (and you could still use `openfl-js`). However an embarrassing bug prevents me to show it for now :D

**Comment by mastef** [comment]
On a positive side-note, I managed to reduce build times with `-Dmodular` from 60 seconds down to 9 seconds by excluding based on top packages, instead of every single `openfl` & `lime` class. https://github.com/mastef/openfl-Dmodular/commit/57ba73f551a6ef680581cc25ed5c02b5e117f528?diff=split - this is a similar approach to @back2dos's comment - thanks for the idea! So now that I was able to build a tiny project in < 10 seconds, need to test now how compile times will scale with more classes ( and if switching to `include` will work better performance wise )

@back2dos A bit off topic but I always run into the same trap with the `--include` and `--exclude` macros after hours of testing... since `include` _only_ takes packages while `exclude` takes class names _or_ packages...

Especially confusing since both take params `pack : String, ?rec = true` and you'd expect the behaviour to match based on the same method signatures. And `strict` in `include` is by default set to `false` - if it would be by default set to `true` you'd be able to catch the inclusion of non-existing parameters much faster ( like testing with class names, which.. just compiles, without having any effect whatsoever ).

I guess what I was hoping for is a macro that allows me to just cherry pick single classes, and not pull any dependencies.

**Comment by mastef** [comment]
I think I tested a sister-project of this, but was annoyed that React was a dependency. But it's an interesting approach to provide multiple .hxml files as webpack entry points. I might take a look into that @elsassph but at this stage it's not yet feasible for me, as the openfl-js haxe externs don't include private methods & properties - and the project I'm working on extends those. But it's getting on my radar now, thank you!

( The other reason is that I might be also testing with Rollup soon, [due to the way Rollup and closure exceed in performance](https://nolanlawson.com/2016/08/15/the-cost-of-small-modules/) based on the amount of included modules, so would not want to lock myself into a webpack based solution atm. )

**Comment by back2dos** [comment]
It's hard to give advise for any project without knowing specifics. But I'm kind of assuming that your dependency graph does not tie everything to everything. If it does, you may also need `--macro exclude`.

Generally I think there are two options, depending on your needs:

1. If you can just prebuild vast parts of the code base and not touch it again, you would build separate JS files (per hxgenjs) with haxe externs (either relying on Joshua's recent work or on `--gen-hx-classes`). Because the results of many small builds are the same as the result of one single build, you can split it up to avoid hitting memory bounds. As said before, skip work for any class that has been generated already, to keep usage lower.
2. If you have to rebuild everything all the time, then add some kind of hash to each generated file (Context.signature might be suitable). Skip work for all classes who's hash still matches the last version. You can fill up this cache with smaller builds until the remaining work is small enough to fit into memory.

Now, whether you use `--macro include` or `--macro exclude` or somehow tell the custom generator to skip certain classes depends on what works the best for you. If the builtin functionality of the compiler does the job, that's great. If not, fiddling with the generator should always get you there ;)
